File: ios-native/Jottery/Jottery/Models/UserSettings.swift
import Foundation
import GRDB

/// User application settings — stored in the `settings` table.
struct UserSettings: Codable, FetchableRecord, PersistableRecord {
    static let databaseTableName = "settings"

    var id: Int = 1
    var language: String
    var theme: String            // "light", "dark", "auto"
    var sortOrder: String        // "recent", "oldest", "alpha", "created"
    var autoLockTimeout: Int     // Minutes
    var syncEnabled: Bool
    var syncEndpoint: String?

    enum CodingKeys: String, CodingKey {
        case id
        case language
        case theme
        case sortOrder = "sort_order"
        case autoLockTimeout = "auto_lock_timeout"
        case syncEnabled = "sync_enabled"
        case syncEndpoint = "sync_endpoint"
    }

    static let defaults = UserSettings(
        language: "system",
        theme: "auto",
        sortOrder: "recent",
        autoLockTimeout: 15,
        syncEnabled: false,
        syncEndpoint: nil
    )

    var themeMode: ThemeMode {
        get { ThemeMode(rawValue: theme) ?? .auto }
        set { theme = newValue.rawValue }
    }

    var sort: SortOrder {
        get { SortOrder(rawValue: sortOrder) ?? .recent }
        set { sortOrder = newValue.rawValue }
    }
}

enum ThemeMode: String, CaseIterable {
    case light, dark, auto
}

enum SortOrder: String, CaseIterable {
    case recent, oldest, alpha, created

    var displayName: String {
        switch self {
        case .recent: return L.sortRecentlyModified
        case .oldest: return L.sortOldestModified
        case .alpha: return L.sortAlphabetical
        case .created: return L.sortDateCreated
        }
    }
}

File: android-native/app/src/main/java/com/jottery/android/model/UserSettings.kt
package com.jottery.android.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey

/**
 * User preferences. Singleton table (id=1).
 */
@Entity(tableName = "settings")
data class UserSettings(
    @PrimaryKey
    val id: Int = 1,

    val language: String = "system",

    val theme: String = "auto",

    @ColumnInfo(name = "sort_order")
    val sortOrder: String = "recent",

    @ColumnInfo(name = "auto_lock_timeout")
    val autoLockTimeout: Int = 15,

    @ColumnInfo(name = "sync_enabled")
    val syncEnabled: Boolean = false,

    @ColumnInfo(name = "sync_endpoint")
    val syncEndpoint: String? = null,

    @ColumnInfo(name = "color_palette")
    val colorPalette: String? = null,

    @ColumnInfo(name = "tag_colors")
    val tagColors: String? = null,
) {
    val themeMode: ThemeMode
        get() = ThemeMode.fromString(theme)

    val sort: SortOrder
        get() = SortOrder.fromString(sortOrder)

    companion object {
        val defaults = UserSettings(
            language = "system",
            theme = "auto",
            sortOrder = "recent",
            autoLockTimeout = 15,
            syncEnabled = false,
            syncEndpoint = null,
        )
    }
}

enum class ThemeMode(val value: String) {
    LIGHT("light"),
    DARK("dark"),
    AUTO("auto");

    companion object {
        fun fromString(value: String): ThemeMode =
            entries.firstOrNull { it.value == value } ?: AUTO
    }
}

enum class SortOrder(val value: String, val displayName: String) {
    RECENT("recent", "Most Recent"),
    OLDEST("oldest", "Oldest First"),
    ALPHA("alpha", "Alphabetical"),
    CREATED("created", "Date Created");

    companion object {
        fun fromString(value: String): SortOrder =
            entries.firstOrNull { it.value == value } ?: RECENT
    }
}
